realm/realm-library/src/main/java/io/realm/BaseRealm.java

```diff
@@ -281,6 +342,41 @@ public void refresh() {
     }
 
     /**
+     * Blocks the current thread util transactions are committed by other threads to this Realm or
+     * {@link #stopWaitForChange()} gets called.
+     *
+     * @return {@code true} if transactions are committed to this Realm, {@code false} if
+     * {@link #stopWaitForChange()} gets called.
+     * @throws IllegalStateException if attempting to wait within a transaction.
+     */
+    public boolean waitForChange() {
+        checkIfValid();
+        if (isInTransaction()) {
+            throw new IllegalStateException("Cannot wait for changes inside of a transaction.");
+        }
+        sharedGroupManager.setWaitForChangeEnabled(true);
+        // add this to wait list for batch stop
+        addRealmToWaitList(this);
+        boolean hasChanged = sharedGroupManager.waitForChange();
+        // remove this from wait list
+        removeRealmFromWaitList(this);
+        if (hasChanged) {
+            sharedGroupManager.advanceRead();
```

Comment:
I think currently the behaviour is actually correct. if user decide to call `waitForChange`, that means user decides to block the current thread and update everything. With async query in queue, there are two cases:
1) async query is not returned. We don't have to update it since it is not returned, and the async results will come as the normal flow.
2) async query has been already done. After `advanceRead`, it should be updated to the latest version.

However, we need to have above in the test case. Changes in the test case need reflect to the `asyncResults1` and `asyncResults2`, and test those two results immediately after `waitForChanges` return.